src/refactorpatterns2/CategorizedDesignPatternInfoImpl.java

```diff
@@ -0,0 +1,100 @@
+ * and open the template in the editor.
+ */
+package refactorpatterns2;
+
+/**
+ *
+ * @author cx
+ */
+class CategorizedDesignPatternInfoImpl extends CategorizedDesignPatternInfo {
+
+    private CategorizedDesignPatternInfoImpl(CategorizedDesignPatternInfoBuilder initializer) {
+        this.patternName = initializer.patternName;
+        this.patternCode = initializer.patternCode;
+        this.patternType = initializer.patternType;
+        this.patternInfo = getDesignPatternInfo(initializer);
+    }
+
+    private String patternName;
+    private String patternCode;
+    private String patternType;
+    private DesignPatternInfo patternInfo;
+
+    @Override
+    public String patternType() {
+        return patternType; // Strings are immutable, so it's ok to return field
+    }
+
+    @Override
+    public String refactorCode(String badCode) {
+        // using patternCode create:
+        return "Refactored code";
+    }
+
+    @Override
+    public String patternName() {
+        return patternName;
+    }
+
+    @Override
+    DesignPatternInfo getDesignPatternInfo() {
+        return new DesignPatternInfoImpl(patternInfo.getUsageExample(), patternInfo.getDescription());
+    }
+
+    private DesignPatternInfo getDesignPatternInfo(CategorizedDesignPatternInfoBuilder initializer) {
+        return new DesignPatternInfoImpl(initializer.patternExample, initializer.patternDescription);
+    }
+
+    public static class CategorizedDesignPatternInfoBuilder {
+
+        private CategorizedDesignPatternInfoBuilder() {
+        }
+
+        public CategorizedDesignPatternInfoBuilder withPatternCode(String patternCode) {
+            this.patternCode = patternCode;
+            return this;
+        }
+
+        public CategorizedDesignPatternInfoBuilder withPatternType(String patternType) {
+            this.patternType = patternType;
+            return this;
+        }
+
+        public CategorizedDesignPatternInfoBuilder withPatternExample(String patternExample) {
+            this.patternExample = patternExample;
+            return this;
+        }
+
+        public CategorizedDesignPatternInfoBuilder withPatternDescription(String patternDescription) {
+            this.patternDescription = patternDescription;
+            return this;
+        }
+
+        public CategorizedDesignPatternInfoBuilder withPatternName(String patternName) {
+            this.patternName = patternName;
+            return this;
+        }
+
+        private String patternName;
+        private String patternCode;
+        private String patternType;
+        private String patternExample;
+        private String patternDescription;
+
+        public CategorizedDesignPatternInfo build() {
+            if (null == patternCode || null == patternType || null == patternName
+                    || null == patternExample || null == patternDescription) {
+                throw new IllegalArgumentException("Arguments must not be null");
```

Comment:
rozumiem idee która stała za tym warunkiem, takie rozwiązanie ma niewątpliwą zaletę. metody ustawiające wartości mogą być wołane w dowolnej kolejności i zawsze widać co jest ustawiane, jednak taki projekt przesuwa proces walidacji do czasu uruchomienia, a z pewnością gdyby udało się osiągnąć walidację w czasie kompilacji byłoby znacznie lepiej. 

Zastanawiam się czy Builder w tym przypadku jest dobrym wyjściem. Z tego co pamiętam Builder sprawdza się najlepiej w sytuacji wielu opcjonalnych argumentów, bo wszystkie obowiązkowe mogą wejść do konstruktora buildera. W tym przypadku wszystkie argumenty należałoby tam umieścić aby uniknąć walidacji a to spowoduje że taka konstrukcja przestanie mieć sens

Reply:
Poprawcie mnie jeśli się mylę, ale nie wiem czy w Javie w ogóle możliwe jest sprawdzenie != null w czasie kompilacji :P 
Zaleta buildera w tym przypadku jest nazwanie parametrów. W przeciwnym wypadku mielibyśmy konstruktor z pięcioma stringami:
`new BuilderPattern("imo","this","deserves","a","builder");`

Reply:
masz rację, trochę mnie poniosło z tą walidacją w czasie kompilacji. nie ma możliwości sprawdzenia stanu zmiennej instancyjnej w czasie kompilacji. 
Miałem jednak na myśli coś w stylu step builder, z przeniesieniem walidacji poza metodę build, tak żeby każda metoda ustawiająca parametr mogła sprawdzić swój argument, a struktura projektu wymuszała zawołanie wszystkich wymaganych metod